### PR TITLE
Add initial specification for a JSON API

### PR DESCRIPTION
**Pull request by arteymix**
In order to provide JSON payloads for the docs, we need a specification. I wrote a simple layout because I think it's worth discussing which and under what name attributes will be delivered.

I did not work into the JSON doclet, but I think it would be worth to reuse its nomenclature.

I also like the idea of JSON-API because there's a lot of relations between symbols and packages and that'll help us structure the information more clearly. I don't think we should adopt the URIs recommendations to preserve a nice and predictable `/{package}/{symbol}` we currently have.

Related PR: https://github.com/Valadoc/valadoc-org/issues/65

### CODE REVIEW

**Comment by astavale** [comment]
I'm curious about the use case for this. From #65 it appears to be a list of available bindings that can be downloaded for offline use. Presumably when the user is online again they can download relevant document packages, like devhelp, for later use. This doesn't seem like a very beneficial use case so it would be useful to learn what further advantages there are.

If the use case is to provide support for IDE tools then I think any work should be aligned with creating a [language server protocol](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md) implementation for Vala. Previously I had thought a `libvala-ide-tools` library would be the best approach because it provides a library with a stable API that then uses an internal API with `libvala`. Now I'm thinking a `vala-language-server` would be more beneficial. The language server protocol is recognised by a number of development environments:

 * [Visual Studio Code](https://code.visualstudio.com/docs/extensions/example-language-server#_explaining-the-client)
 * [Eclipse](https://projects.eclipse.org/projects/technology.lsp4e)
 * [GNOME Builder](https://git.gnome.org/browse/gnome-builder/tree/libide/langserv)
 * [Atom](https://atom.io/packages/atom-languageclient)
 * [Vim](https://github.com/prabirshrestha/vim-lsp)
 * [Emacs](https://github.com/emacs-lsp/lsp-mode/)

For more information, including languages supported, see [Protocol Implementations](https://github.com/Microsoft/language-server-protocol/wiki/Protocol-Implementations).

The three most relevant methods of the language server protocol for a documentation provider are:

 1. [Completion request](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#textDocument_completion)
 2. [Hover Request](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#textDocument_hover)
 3. [Signature help request](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#textDocument_signatureHelp)

All of these methods require a position in the document to be passed and that doesn't make sense for a valadoc.org API, instead a string would be passed. The methods, however, might provide some insights for an API. Also, I think it is a good idea to align any changes to valadoc needed to support a valadoc.org API to match the language server protocol methods. Note that valadoc has been [merged in to the Vala source tree for Vala 0.38](https://bugzilla.gnome.org/show_bug.cgi?id=782782). Some examples would be:

 * `textDocument/completion` returns `CompletionItem`s that can include a documentation string and the kind of item, e.g. method, variable, etc. So GET /{package}/{symbol} in the valadoc.org API may need a field for the kind of symbol returned
 * `textDocument/hover` returns a list of `MarkedString`s that are either Markdown documentation or code snippets with the language identified
 * `textDocument/signatureHelp` returns a list of `SignatureInformation` each containing a list of `ParameterInformation`. Is that level of detail out of scope for a valadoc.org API?

The language server protocol is transport agnostic and implementations use STDIN/STDOUT, named pipes and TCP. Is a valadoc.org API there to support a language server for remote lookups over the network or to help progress `valadoc` to the point it can support a locally running `vala-language-server` by parsing local VAPIs/GIRs etc.?

The language server protocol uses JSON-RPC v2. This is probably not relevant to a valadoc.org API because it would mean the API supporting individual editing sessions, which would be beyond the resources available from the server.

In summary I think:
 * the use cases for a valadoc.org API need to be made explicit
 * the use cases would then guide the extent of the API, e.g. should signature/parameter information be part of the API?
 * work should be aligned with a possible implementation of a Vala language server

**Comment by arteymix** [comment]
The main goal of having a Valadoc API would be to provide an easy-to-consume representation of the current HTML-generated documentation.

The use cases would be quite wide:

 - rewrite the current client code of Valadoc.org around the standardized search API and extract tooptips from well-structured structured information
 - automatic linking of `@see`-like annotations (see http://jsonapi.org/format/#document-links)
 - mobile application for consulting the docs
 - integrate symbol documentation in a future language server

The idea I have behind the implementation of this "service" is to write a valadoc doclet that would emit a hiearchy with hyper-links that could reside on https://valadoc.org/ or locally. It would surely go upstream when ready!

Ideally, a language server would parse documentation directly from the sources, but it's not always available from VAPIs and GIRs. Thus, I think that querying valadoc.org would provide a good alternative.

**Comment by benwaffle** [comment]
I doubt we want to write a language server that calls out to valadoc.org - it would be very slow. It would be nice to have a valadoc.org client in GNOME Builder though.

FYI i wrote an XML generator doclet here: https://github.com/eustasy/vala-docs.org

File: valadoc-spec.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  description: >
+    The Valadoc.org API specifies how documentation is provided and structured
+    in the in JSON format.
+host: valadoc.org
+basePath: /
+schemes:
+  - http
+  - https
+produces:
+  - application/vnd.api+json
```
Comment (benwaffle) [inline valadoc-spec.yml]: json mime type is `application/json`, no?

Comment (lewisgoddard) [inline valadoc-spec.yml]: Not when it's a specified API http://jsonapi.org/

Comment (arteymix) [inline valadoc-spec.yml]: We can do both using content negotiation.

**Comment by ZanderBrown** [comment]
Any progress on this?

**Comment by tintou** [comment]
No progress since a year, closing now
